File: packages/go/cost.go
package openrouterauto

import (
	"encoding/json"
	"strconv"
	"strings"

	_ "embed"
)

//go:embed registry/cost.json
var costJSON []byte

type priceTierEntry struct {
	MaxAvgPrice *float64 `json:"max_avg_price"`
}

type costRegistry struct {
	PriceTiers                 map[string]priceTierEntry `json:"price_tiers"`
	TokenEstimateCharsPerToken int                       `json:"token_estimate_chars_per_token"`
	MessageOverheadTokens      int                       `json:"message_overhead_tokens"`
}

var costReg costRegistry

func init() {
	if err := json.Unmarshal(costJSON, &costReg); err != nil {
		panic("openrouterauto: failed to load cost.json: " + err.Error())
	}
}

// IsFreeModel returns true when the model carries zero-cost pricing or a :free suffix.
func IsFreeModel(model Model) bool {
	if strings.HasSuffix(model.ID, ":free") {
		return true
	}
	return parsePrice(model.Pricing.Prompt) == 0 && parsePrice(model.Pricing.Completion) == 0
}

// GetPriceTier classifies a model as "free", "cheap", "moderate", or "expensive".
func GetPriceTier(model Model) string {
	if IsFreeModel(model) {
		return "free"
	}
	avg := (parsePrice(model.Pricing.Prompt) + parsePrice(model.Pricing.Completion)) / 2
	for _, tier := range []string{"cheap", "moderate"} {
		entry, ok := costReg.PriceTiers[tier]
		if ok && entry.MaxAvgPrice != nil && avg <= *entry.MaxAvgPrice {
			return tier
		}
	}
	return "expensive"
}

// CalculateCost returns a CostEstimate for the given token counts.
// The OpenRouter API returns prices as per-1K-token strings.
func CalculateCost(model Model, promptTokens, completionTokens int) CostEstimate {
	pp := parsePrice(model.Pricing.Prompt)
	cp := parsePrice(model.Pricing.Completion)
	promptCost := (float64(promptTokens) / 1000) * pp
	completionCost := (float64(completionTokens) / 1000) * cp
	return CostEstimate{
		PromptCost:     promptCost,
		CompletionCost: completionCost,
		TotalCost:      promptCost + completionCost,
		Currency:       "USD",
	}
}

// EstimateTokens returns a rough token count for text using the registry ratio.
func EstimateTokens(text string) int {
	if text == "" {
		return 0
	}
	cpt := costReg.TokenEstimateCharsPerToken
	if cpt <= 0 {
		cpt = 4
	}
	return (len(text) + cpt - 1) / cpt
}

// parsePrice converts the OpenRouter string price to float64.
func parsePrice(s string) float64 {
	if s == "" {
		return 0
	}
	f, _ := strconv.ParseFloat(s, 64)
	return f
}

File: packages/rust/src/cost.rs
use crate::types::{CostEstimate, Model};
use once_cell::sync::Lazy;
use serde::Deserialize;
use std::collections::HashMap;

const COST_JSON: &str = include_str!("../registry/cost.json");

#[derive(Deserialize)]
struct PriceTierEntry {
    max_avg_price: Option<f64>,
}

#[derive(Deserialize)]
struct CostRegistry {
    price_tiers: HashMap<String, PriceTierEntry>,
    token_estimate_chars_per_token: usize,
    #[allow(dead_code)]
    message_overhead_tokens: usize,
}

static COST_REG: Lazy<CostRegistry> =
    Lazy::new(|| serde_json::from_str(COST_JSON).expect("failed to parse cost.json"));

/// Returns true when the model carries zero-cost pricing or a `:free` suffix.
pub fn is_free_model(model: &Model) -> bool {
    if model.id.ends_with(":free") {
        return true;
    }
    parse_price(&model.pricing.prompt) == 0.0 && parse_price(&model.pricing.completion) == 0.0
}

/// Classify a model as "free", "cheap", "moderate", or "expensive".
pub fn get_price_tier(model: &Model) -> &'static str {
    if is_free_model(model) {
        return "free";
    }
    let avg = (parse_price(&model.pricing.prompt) + parse_price(&model.pricing.completion)) / 2.0;
    for tier in &["cheap", "moderate"] {
        if let Some(entry) = COST_REG.price_tiers.get(*tier) {
            if let Some(max) = entry.max_avg_price {
                if avg <= max {
                    return tier;
                }
            }
        }
    }
    "expensive"
}

/// Calculate cost for the given token counts.
pub fn calculate_cost(model: &Model, prompt_tokens: i64, completion_tokens: i64) -> CostEstimate {
    let pp = parse_price(&model.pricing.prompt);
    let cp = parse_price(&model.pricing.completion);
    let prompt_cost = (prompt_tokens as f64 / 1000.0) * pp;
    let completion_cost = (completion_tokens as f64 / 1000.0) * cp;
    CostEstimate {
        prompt_cost,
        completion_cost,
        total_cost: prompt_cost + completion_cost,
        currency: "USD".to_string(),
    }
}

/// Rough token count using the registry chars-per-token ratio.
pub fn estimate_tokens(text: &str) -> usize {
    if text.is_empty() {
        return 0;
    }
    let cpt = COST_REG.token_estimate_chars_per_token.max(1);
    (text.len() + cpt - 1) / cpt
}

fn parse_price(s: &str) -> f64 {
    s.parse::<f64>().unwrap_or(0.0)
}
